implement polygon query

rogerkuou opened the pull request:
Implement polygon enrichment and polygon subset.

rogerkuou commented:
Hi @meiertgrootes. I implemented the polygon query in this branch. Could you please review it?

You can use [this notebook](https://github.com/MotionbyLearning/stm/blob/3_query_from_polygon/examples/demo_stm.ipynb) to test.

Your example notebook has been moved [here](https://github.com/MotionbyLearning/stm/blob/3_query_from_polygon/examples/legacy/polygon_demo.ipynb).

rogerkuou commented:
@meiertgrootes I have updated this PR with new example notebooks, on both local and Spider. Thanks for the reviewing!

meiertgrootes commented on stm/stm.py:
```diff
@@ -0,0 +1,418 @@
+            A subset of the original STM.
+        """
+
+        match method:  # Match statements available only from python 3.10 onwards
```
I still think too much of each subset routine's base code is within the routine. I would suggest trying to move that out, only making a call from the subset function. This would also help with future additions  of subset methods

rogerkuou replied:
converted to #16

meiertgrootes commented on stm/stm.py:
```diff
@@ -0,0 +1,418 @@
+                    **kwargs
+                )  # Check is all required kwargs are available
+                if kwargs["threshold"][0] == "<":
+                    str_parts = kwargs["threshold"].partition("<")
```
only ">" or "<" are possible. Users may expect to also be able to use "<=" or "=>". Can this be supported?

A thought for the future might be a band/range selection

rogerkuou replied:
converted to #17

meiertgrootes commented on stm/stm.py:
```diff
@@ -0,0 +1,418 @@
+                    )
+            case "density":
+                check_density_kwargs(**kwargs)  # Check for all require kwargs
+                gdf = gpd.GeoDataFrame(
```
The implementation of the 'density' subset method is elegant. A bit more in-line documentation of the algorithm would help the user immediately understand the logic.

rogerkuou replied:
Thx, I added some in-line docs

meiertgrootes commented on stm/stm.py:
```diff
@@ -0,0 +1,418 @@
+                    (max(self._obj[kwargs["y"]]) - min(self._obj[kwargs["y"]]))
+                    / kwargs["dy"]
+                )
+                raster = features.rasterize(
```
A possible draw back is having to create ever finer rasters for higher desired densities, as well as the need for possibly non-intuuitive cell sizes.

E.g: select at density 1/m^2 -> grid of dx = dy = 1m
how to select at e.g 5/m^2

rogerkuou replied:
Thx for the heads up Meiert! For this function, the application is only for select points to make a network, and there should not be too short arcs within the network. Then I think the requirement is to ensure there is no second point within a specific range. Therefore I think the current implementation is fine.

meiertgrootes commented on stm/stm.py:
```diff
@@ -0,0 +1,418 @@
+            raise ValueError("fields need to be a Iterable or a string")
+
+        # Get polygon type and the first row
+        if isinstance(polygon, gpd.GeoDataFrame):
```
adding support for WKT polygons (as strings) might be useful for users here

rogerkuou replied:
converted to #18

meiertgrootes commented on stm/stm.py:
```diff
@@ -0,0 +1,418 @@
+    return _ds
+
+
+def _ml_str_query(xx, yy, polygon, type_polygon):
```
this is currently based on the "contains" predicate but others are possible in principle. Maybe consider exposing the chosen predicate here via keyword, for possible use in other fct.

meiertgrootes reviewed (approved):
Looks good to merge to me. I do have a number of comments, however. These are more about some possible refactoring and expansion than any blocking issues. Please consider converting some into issues so we can pick them up later while still adding the basic functionality to the main branch now

rogerkuou commented:
Thx @meiertgrootes for the feedback! I agree with almost all of them. I converted most of your comments to issues. Also added in-line documentation for density subsetting.

Just [one thought](https://github.com/MotionbyLearning/stm/pull/10#discussion_r1119929373) on the density implementation.

Will merge it for now
